kubernetes-discovery-client/src/main/java/io/micronaut/kubernetes/discovery/provider/AbstractV1EndpointsProvider.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.kubernetes.discovery.provider;
+
+import io.kubernetes.client.openapi.models.V1Endpoints;
+import io.kubernetes.client.openapi.models.V1ObjectMeta;
+import io.micronaut.discovery.ServiceInstance;
+import io.micronaut.kubernetes.KubernetesConfiguration;
+import io.micronaut.kubernetes.discovery.AbstractKubernetesServiceInstanceProvider;
+import io.micronaut.kubernetes.discovery.KubernetesServiceConfiguration;
+import io.micronaut.kubernetes.util.KubernetesUtils;
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Service instance provider uses Kubernetes Endpoints as source of service discovery.
+ *
+ * @author Pavol Gressa
+ * @since 3.2
+ */
+public abstract class AbstractV1EndpointsProvider extends AbstractKubernetesServiceInstanceProvider {
```

Comment:
This abstract class contains the filtering and transformation of V1Endpoint to the service materials needed by DiscoveryClient. Afterwars the implementation just specifies the method how are the V1Endpoint obtained. At this moment is't either by directly fetching the resource from K8s API or using informer.